python/pyspark/ml/functions.py

```diff
@@ -106,6 +112,170 @@ def array_to_vector(col: Column) -> Column:
+    """
+    # generate a new uuid each time this is invoked on the driver to invalidate executor-side cache.
+    model_uuid = uuid.uuid4()
+
+    def predict(data: Iterator[pd.DataFrame]) -> Iterator[pd.DataFrame]:
+        import pyspark.ml.executor_globals as exec_global
+
+        if exec_global.predict_fn and exec_global.model_uuid == model_uuid:
+            predict_fn = exec_global.predict_fn
+        else:
+            predict_fn = predict_batch_fn(**kwargs)
+            exec_global.predict_fn = predict_fn
+            exec_global.model_uuid = model_uuid
+
+        for partition in data:
+            has_tensors = has_tensor_cols(partition)
+            for batch in batched(partition, batch_size):
+                inputs: Union[np.ndarray, dict[str, np.ndarray]]
+                if input_names:
+                    # input names provided, expect a dictionary of named numpy arrays
+                    # check if the number of inputs matches expected
+                    num_expected = len(input_names)
+                    num_actual = len(batch.columns)
+                    if num_actual != num_expected:
+                        msg = "Model expected {} inputs, but received {} columns"
+                        raise ValueError(msg.format(num_expected, num_actual))
+
+                    # rename dataframe column names to match model input names, if needed
+                    if input_names != list(batch.columns):
+                        batch.columns = input_names
+
+                    if has_tensors:
+                        raise ValueError("Tensor columns require an input_tensor_shape")
+
+                    # create a dictionary of named inputs
+                    inputs_dict = batch.to_dict(orient="series")
+
+                    # reshape inputs, if needed
+                    if input_tensor_shapes:
+                        if len(input_tensor_shapes) == num_actual:
+                            for i, (k, v) in enumerate(inputs_dict.items()):
+                                inputs_dict[k] = v.reshape(input_tensor_shapes[i])  # type: ignore
+                        else:
+                            raise ValueError("input_tensor_shapes must match columns")
+
+                    inputs = inputs_dict  # type: ignore
+                else:
+                    # no input names provided, expect a single numpy array
+                    if input_tensor_shapes:
+                        if len(input_tensor_shapes) == 1:
+                            if len(batch.columns) == 1:
+                                # if one tensor input and one column, vstack and reshape the batch
+                                input_shape = input_tensor_shapes[0]
+                                input_shape[0] = -1  # replace None with -1 in batch dimension
+                                # input = np.vstack(batch).reshape(input_shape)     # name, col
+                                inputs = np.vstack(batch.iloc[:, 0]).reshape(input_shape)  # struct
```

Comment:
So the `input_shape` (comes from `input_tensor_shapes` argument)  must be [-1, xx, ..], first dimension should always be `-1`.
I think we can let user pass shape arguments without the first `-1`

Reply:
Was mostly trying to follow (what I thought was) standard practice, which is also commonly used for `np.reshape` operations, s.t. the end-user code looks similar to single-node code, but I'm open to changing this...

Reply:
Went ahead and removed the requirement to pass the -1 batch dim.  Also, added support for "sparse" representation of the `input_tensor_shapes`, per offline suggestion from @mengxr.